File: packages/core/src/commands/info.ts
import type { Content, File } from '../content.ts'
import { shortcuts } from '../content.ts'
import type { Line, Output } from '../output.ts'
import { blankLine, plainLine, styledLine } from '../output.ts'
import type { Session } from '../session.ts'
import { lines } from '../text.ts'

// Neutral voice, intentionally. A journey can be someone's or a company's.
// "What I build with" or "leave me a message" ring false when spoken by a company.

/** Moving around. Same in every journey. */
const NAVIGATION: ReadonlyArray<readonly [string, string]> = [
  ['ls', 'list what is here, -a shows everything'],
  ['cd', 'move around, .. goes up'],
  ['pwd', 'where you are right now'],
  ['cat', 'read a file'],
]

/** Reaching out and running the session. Same in every journey. */
const SESSION: ReadonlyArray<readonly [string, string]> = [
  ['contact', 'leave a message right here'],
  ['history', 'what you have typed'],
  ['clear', 'wipe the screen'],
  ['exit', 'close the session'],
]

/**
 * Grouped help, with an example per group. A flat wall of commands is a scan
 * problem; three short labeled blocks tell a new visitor what exists and,
 * with one example each, what the shape of a command is (clig.dev: nudge,
 * show examples, keep it scannable).
 */
export function help(c: Content): Output {
  const journeySections = shortcuts(c).map((f) => [f.name, f.title] as const)
  const groups: ReadonlyArray<readonly [string, ReadonlyArray<readonly [string, string]>]> = [
    ['navigate', NAVIGATION],
    ['this journey', journeySections],
    ['reach out & session', SESSION],
  ]

  const all = groups.flatMap(([, rows]) => rows)
  const width = Math.max(9, ...all.map(([name]) => name.length + 2))

  const out: Line[] = [styledLine('commands', 'bold'), blankLine()]
  for (const [heading, rows] of groups) {
    if (rows.length > 0) out.push(styledLine(`  ${heading}`, 'bold'))
    for (const [name, description] of rows) {
      out.push({
        spans: [
          { text: `    ${name.padEnd(width)}`, style: 'accent' },
          { text: description, style: 'dim' },
        ],
      })
    }
  }
  out.push(blankLine())
  out.push({ spans: [{ text: '    example: cat <file> reads a file', style: 'dim' }] })
  return { lines: out, failed: false }
}

export function history(s: Session): Output {
  return {
    // Rust's `format!("{:>3}  {entry}")`: number right-aligned to 3, then two spaces.
    lines: s.history.map((entry, i) => plainLine(`${String(i + 1).padStart(3)}  ${entry}`)),
    failed: false,
  }
}

export function show(file: File): Output {
  return { lines: lines(file.body).map(plainLine), failed: false }
}

File: crates/core/src/commands/info.rs
use crate::content::Content;
use crate::output::{Line, Output, Span, Style};
use crate::session::Session;

const COMMANDS: &[(&str, &str)] = &[
    ("ls", "list what is here, -a shows everything"),
    ("cd", "move around, .. goes up"),
    ("pwd", "where you are right now"),
    ("cat", "read a file"),
    ("whoami", "the short version"),
    ("stack", "what I build with"),
    ("cv", "the PDF, for your ATS"),
    ("contact", "leave me a message right here"),
    ("book", "put something in the calendar"),
    ("history", "what you have typed"),
    ("clear", "wipe the screen"),
    ("exit", "close the session"),
];

pub fn help() -> Output {
    let mut lines = vec![Line::styled("commands", Style::Bold), Line::blank()];
    for (name, description) in COMMANDS {
        lines.push(Line {
            spans: vec![
                Span {
                    text: format!("  {name:<9}"),
                    style: Style::Accent,
                },
                Span {
                    text: description.to_string(),
                    style: Style::Dim,
                },
            ],
        });
    }
    Output {
        lines,
        ..Output::default()
    }
}

pub fn history(session: &Session) -> Output {
    Output {
        lines: session
            .history
            .iter()
            .enumerate()
            .map(|(i, entry)| Line::plain(&format!("{:>3}  {entry}", i + 1)))
            .collect(),
        ..Output::default()
    }
}

pub fn show(content: &Content, name: &str) -> Output {
    match content.resolve_file(&[name.to_string()]) {
        Some(file) => Output {
            lines: file.body.lines().map(Line::plain).collect(),
            ..Output::default()
        },
        None => Output::failure(&format!("{name}: content is missing")),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::session::Session;

    fn flatten(out: &Output) -> String {
        out.lines
            .iter()
            .map(|l| l.spans.iter().map(|s| s.text.as_str()).collect::<String>())
            .collect::<Vec<_>>()
            .join("\n")
    }

    #[test]
    fn help_lists_the_main_commands() {
        let rendered = flatten(&help());
        for name in [
            "ls", "cd", "pwd", "cat", "whoami", "stack", "cv", "contact", "book", "exit",
        ] {
            assert!(rendered.contains(name), "help is missing {name}");
        }
    }

    #[test]
    fn help_never_mentions_the_egg() {
        assert!(!flatten(&help()).contains("knock"));
    }

    #[test]
    fn history_numbers_the_lines() {
        let mut session = Session::new();
        session.history.push("ls".to_string());
        session.history.push("whoami".to_string());
        let rendered = flatten(&history(&session));
        assert!(rendered.contains("1  ls"));
        assert!(rendered.contains("2  whoami"));
    }
}
